scripts/asr_language_modelling/ngram_lm/eval_beamsearch_ngram.py

```diff
@@ -0,0 +1,186 @@
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='Evaluate an ASR model with beam search decoding and an n-gram KenLM language model.'
+    )
+    parser.add_argument("--nemo_model_file", required=True, type=str)
+    parser.add_argument("--kenlm_model_file", required=False, default=None, type=str)
+    parser.add_argument("--input_manifest", required=True, type=str)
+    parser.add_argument("--preds_output_file", required=True, type=str)
+    parser.add_argument("--probs_cache_file", default=None, type=str)
+    parser.add_argument("--use_probs_cache", action="store_true")
+    parser.add_argument("--beam_width", required=True, type=int)
+    parser.add_argument("--beam_alpha", required=True, type=float)
+    parser.add_argument("--beam_beta", required=True, type=float)
+    parser.add_argument("--acoustic_batch_size", default=16, type=int)
+    parser.add_argument("--beam_batch_size", default=128, type=int)
+    parser.add_argument("--device", default="cuda:0", type=str)
+    parser.add_argument(
+        "--decoding_mode", choices=["greedy", "beamsearch", "beamsearch_ngram"], default="beamsearch_ngram", type=str
+    )
+    args = parser.parse_args()
+
+    logging.info(f"BEAM WIDTH : {args.beam_width}")
+    logging.info(f"BEAM ALPHA : {args.beam_alpha}")
+    logging.info(f"BEAM BETA : {args.beam_beta}")
+
+    asr_model = nemo_asr.models.EncDecCTCModelBPE.restore_from(
+        args.nemo_model_file, map_location=torch.device(args.device)
+    )
+    asr_model.preprocessor.featurizer.dither = 0
+    asr_model.preprocessor.featurizer.pad_to = 0
+    # Set model to inference mode
+    asr_model.eval()
+
+    model_tokenizer = asr_model.tokenizer
+    vocabs = list(model_tokenizer.tokenizer.get_vocab().keys())
+
+    target_transcripts = []
+    with open(args.input_manifest, 'r') as manifest_file:
+        audio_file_paths = []
+        for line in tqdm(manifest_file, desc=f"Reading Manifest {args.input_manifest} ..."):
+            data = json.loads(line)
+            target_transcripts.append(data['text'])
+            audio_file_paths.append(data['audio_filepath'])
+
+    # drop it later
+    # audio_file_paths = audio_file_paths[0:10]
+
+    if args.use_probs_cache and os.path.exists(args.probs_cache_file):
+        logging.info(f"Found a pickle file of probabilities at {args.probs_cache_file}.")
+        logging.info(f"Loading the cached pickle file of probabilities from {args.probs_cache_file}...")
+        with open(args.probs_cache_file, 'rb') as probs_file:
+            all_probs = pickle.load(probs_file)
+
+        if len(all_probs) != len(audio_file_paths):
+            raise ValueError(
+                f"The number of samples in the probabilities file '{args.probs_cache_file}' is not "
+                f"the same as the manifest file. "
+                f"You may need to delete the probabilities cached file."
+            )
+    else:
+        with torch.no_grad():
```

Comment:
We can use amp if its available? Would speed up decoding a lot for larger models. Theres a way to setup a context_manager which checks for amp or no amp and switches inside the `transcribe_speech.py` example.

Reply:
Added the feature to use amp.

Reply:
Added the support for amp.